src/fhir.js

```diff
@@ -290,35 +291,63 @@ class FHIRObject {
 }
 
 class Patient extends FHIRObject {
-  constructor(bundle, modelInfo) {
+  constructor(bundle, modelInfo, shouldCheckProfile = false) {
     const patientClass = modelInfo.patientClassIdentifier
       ? modelInfo.patientClassIdentifier
       : modelInfo.patientClassName;
     const resourceType = modelInfo.patientClassName.replace(/^FHIR\./, '');
     const ptEntry = bundle.entry.find(e => e.resource && e.resource.resourceType == resourceType);
     const ptClass = modelInfo.findClass(patientClass);
     super(ptEntry.resource, ptClass, modelInfo);
+    this._shouldCheckProfile = shouldCheckProfile;
+
     // Define a "private" un-enumerable property to hold the bundle
     Object.defineProperty(this, '_bundle', { value: bundle, enumerable: false });
   }
 
-  findRecord(profile) {
-    const records = this.findRecords(profile);
+  findRecord(profile, retrieveDetails) {
+    const records = this.findRecords(profile, retrieveDetails);
     if (records.length > 0) {
       return records[0];
     }
   }
 
-  findRecords(profile) {
-    const classInfo = this._modelInfo.findClass(profile);
+  findRecords(profile, retrieveDetails) {
+    // retrieveDetails was introduced in cql-execution v2.4.1. If it is missing from the function call
+    // profile checking will not work,
+    if (this._shouldCheckProfile && retrieveDetails == null) {
+      throw new Error(
+        'meta.profile checking is only supported using cql-execution >=2.4.1. Please upgrade or disable usage of meta.profile checking'
+      );
+    }
+
+    // Preferring the datatype on retrieveDetails allows the engine to properly identify resources where
+    // the ELM uses a profile from a specific IG (e.g. US Core)
+    const classInfo = this._modelInfo.findClass(
+      retrieveDetails ? retrieveDetails.datatype : profile
+    );
+
     if (classInfo == null) {
       console.error(`Failed to find type info for ${profile}`);
       return [];
     }
     const resourceType = classInfo.name.replace(/^FHIR\./, '');
     const records = this._bundle.entry
       .filter(e => {
-        return e.resource && e.resource.resourceType == resourceType;
+        if (e.resource && e.resource.resourceType == resourceType) {
+          if (
+            this._shouldCheckProfile &&
+            profile !== `http://hl7.org/fhir/StructureDefinition/${resourceType}`
+          ) {
+            return (
+              e.resource.meta &&
+              e.resource.meta.profile &&
+              e.resource.meta.profile.includes(profile)
```

Comment:
It's tempting to upgrade to Node 18 just so you can say `e.resource.meta?.profile?.includes(profile)`. ;-)

Reply:
bleh :/